compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64MacroAssembler.java

```diff
@@ -773,4 +777,412 @@ public void stringIndexOf(Register str1, Register str2,
 
     }
 
+    /* Compress a UTF16 string which de facto is a Latin1 string into a byte array
+     * representation (buffer).
+     *
+     *   src   (rsi) the start address of source char[] to be compressed
+     *   dst   (rdi) the start address of destination byte[] vector
+     *   len   (rdx) the length
+     *   vtmp1 (xmm)
+     *   vtmp2 (xmm)
+     *   vtmp3 (xmm)
+     *   vtmp4 (xmm)
+     *   tmp   (gpr)
+     *   res   (rax) the result code (length on success, zero otherwise)
+     */
+    public void char_array_compress(Register src, Register dst, Register len,
```

Comment:
Maybe I'm alone with my opinion but I don't see a reason why having a method for these in MacroAssembler is better than just having the code in the Op. All the ops I've added in the past just have the assembler code in them. Makes it somewhat easier to read and understand.

Reply:
Essentially I'm just following suite with *AMD64StringIndexOfOp*. However, I would argue that providing the implementation of two rather large and similar/related assembly routines are best located together instead of spread out in two not obviously related files, interleaved with what is boilerplate only. This way, the assembly routines read very much like regular assembly code (a suitable idiom) and the boilerplate is kept in the "Op" classes. I think it makes for considerably easier reading of the assembly while keeping the boilerplate code at a "follow-the-template" level.

Reply:
Look at `AMD64ArrayEqualsOp` instead.

You could argue that having similar assembly code in two different files makes it easier to look at/compare because you could open the two files side-by-side :-)  Weak argument, I know...